File: app/src/main/java/info/nightscout/androidaps/utils/OKDialog.java
package info.nightscout.androidaps.utils;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.os.Handler;
import android.os.SystemClock;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.view.ContextThemeWrapper;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;

public class OKDialog {
    public static void show(final Context context, String title, String message, final Runnable runnable) {
        if (title.isEmpty()) title = MainApp.gs(R.string.message);
        View titleLayout = LayoutInflater.from(context).inflate(R.layout.dialog_alert_custom, null);
        ((TextView)titleLayout.findViewById(R.id.alertdialog_title)).setText(title);
        ((ImageView)titleLayout.findViewById(R.id.alertdialog_icon)).setImageResource(R.drawable.ic_check_while_48dp);

        new AlertDialog.Builder(new ContextThemeWrapper(context, R.style.AppTheme))
                .setCustomTitle(titleLayout)
                .setMessage(message)
                .setPositiveButton(MainApp.gs(R.string.ok), (dialog, which) -> {
                    dialog.dismiss();
                    if (runnable != null) {
                        SystemClock.sleep(100);
                        runOnUiThread(runnable);
                    }
                })
                .show()
                .setCanceledOnTouchOutside(false);
    }

    public static boolean runOnUiThread(Runnable theRunnable) {
        final Handler mainHandler = new Handler(MainApp.instance().getApplicationContext().getMainLooper());
        return mainHandler.post(theRunnable);
    }

    public static void show(final Activity activity, String title, Spanned message, final Runnable runnable) {
        if (title.isEmpty()) title = MainApp.gs(R.string.message);
        View titleLayout = activity.getLayoutInflater().inflate(R.layout.dialog_alert_custom, null);
        ((TextView)titleLayout.findViewById(R.id.alertdialog_title)).setText(title);
        ((ImageView)titleLayout.findViewById(R.id.alertdialog_icon)).setImageResource(R.drawable.ic_check_while_48dp);

        new AlertDialog.Builder(new ContextThemeWrapper(activity, R.style.AppTheme))
                .setCustomTitle(titleLayout)
                .setMessage(message)
                .setPositiveButton(MainApp.gs(R.string.ok), (dialog, which) -> {
                    dialog.dismiss();
                    if (runnable != null) {
                        SystemClock.sleep(100);
                        activity.runOnUiThread(runnable);
                    }
                })
                .show()
                .setCanceledOnTouchOutside(false);
    }

    public static void showConfirmation(final Activity activity, String message, final Runnable ok) {
        showConfirmation(activity, message, ok, null);
    }

    public static void showConfirmation(final Activity activity, Spanned message, final Runnable ok) {
        showConfirmation(activity, message, ok, null);
    }

    public static void showConfirmation(final Activity activity, Spanned message, final Runnable ok, final Runnable cancel) {
        View titleLayout = activity.getLayoutInflater().inflate(R.layout.dialog_alert_custom, null);
        ((TextView)titleLayout.findViewById(R.id.alertdialog_title)).setText(R.string.confirmation);
        ((ImageView)titleLayout.findViewById(R.id.alertdialog_icon)).setImageResource(R.drawable.ic_check_while_48dp);

        new AlertDialog.Builder(new ContextThemeWrapper(activity, R.style.AppTheme))
                .setMessage(message)
                .setCustomTitle(titleLayout)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    dialog.dismiss();
                    if (ok != null) {
                        SystemClock.sleep(100);
                        activity.runOnUiThread(ok);
                    }
                })
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> {
                    dialog.dismiss();
                    if (cancel != null) {
                        SystemClock.sleep(100);
                        activity.runOnUiThread(cancel);
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show()
                .setCanceledOnTouchOutside(false);
    }

    public static void showConfirmation(final Activity activity, String message, final Runnable ok, final Runnable cancel) {
        View titleLayout = activity.getLayoutInflater().inflate(R.layout.dialog_alert_custom, null);
        ((TextView)titleLayout.findViewById(R.id.alertdialog_title)).setText(R.string.confirmation);
        ((ImageView)titleLayout.findViewById(R.id.alertdialog_icon)).setImageResource(R.drawable.ic_check_while_48dp);

        new AlertDialog.Builder(new ContextThemeWrapper(activity, R.style.AppTheme))
                .setMessage(message)
                .setCustomTitle(titleLayout)
                .setView(R.layout.dialog_alert_custom)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    dialog.dismiss();
                    if (ok != null) {
                        SystemClock.sleep(100);
                        activity.runOnUiThread(ok);
                    }
                })
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> {
                    dialog.dismiss();
                    if (cancel != null) {
                        SystemClock.sleep(100);
                        activity.runOnUiThread(cancel);
                    }
                })
                .show()
                .setCanceledOnTouchOutside(false);
    }

}

File: app/src/main/java/info/nightscout/androidaps/utils/OKDialog.kt
package info.nightscout.androidaps.utils

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.os.Handler
import android.os.SystemClock
import android.text.Spanned
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.view.ContextThemeWrapper
import info.nightscout.androidaps.MainApp
import info.nightscout.androidaps.R

object OKDialog {
    @JvmStatic
    @JvmOverloads
    fun show(context: Context, title: String, message: String, runnable: Runnable? = null) {
        var notEmptytitle = title
        if (notEmptytitle.isEmpty()) notEmptytitle = MainApp.gs(R.string.message)
        val titleLayout = LayoutInflater.from(context).inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = notEmptytitle
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(context, R.style.AppTheme))
            .setCustomTitle(titleLayout)
            .setMessage(message)
            .setPositiveButton(MainApp.gs(R.string.ok)) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                runOnUiThread(runnable)
            }
            .show()
            .setCanceledOnTouchOutside(false)
    }

    fun runOnUiThread(theRunnable: Runnable?) {
        val mainHandler = Handler(MainApp.instance().applicationContext.mainLooper)
        theRunnable?.let { mainHandler.post(it) }
    }

    @JvmStatic
    @JvmOverloads
    fun show(activity: Activity, title: String, message: Spanned, runnable: Runnable? = null) {
        var notEmptytitle = title
        if (notEmptytitle.isEmpty()) notEmptytitle = MainApp.gs(R.string.message)
        val titleLayout = activity.layoutInflater.inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = notEmptytitle
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(activity, R.style.AppTheme))
            .setCustomTitle(titleLayout)
            .setMessage(message)
            .setPositiveButton(MainApp.gs(R.string.ok)) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                runnable?.let { activity.runOnUiThread(it) }
            }
            .show()
            .setCanceledOnTouchOutside(false)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(activity: Activity, message: String, ok: Runnable?) {
        showConfirmation(activity, MainApp.gs(R.string.confirmation), message, ok, null)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(activity: Activity, message: Spanned, ok: Runnable?) {
        showConfirmation(activity, MainApp.gs(R.string.confirmation), message, ok, null)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(activity: Activity, title: String, message: Spanned, ok: Runnable?, cancel: Runnable? = null) {
        val titleLayout = activity.layoutInflater.inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = title
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(activity, R.style.AppTheme))
            .setMessage(message)
            .setCustomTitle(titleLayout)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                ok?.let { activity.runOnUiThread(it) }
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                cancel?.let { activity.runOnUiThread(it) }
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
            .setCanceledOnTouchOutside(false)
    }

    @JvmStatic
    fun showConfirmation(activity: Activity, title: String, message: String, ok: Runnable?, cancel: Runnable? = null) {
        val titleLayout = activity.layoutInflater.inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = title
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(activity, R.style.AppTheme))
            .setMessage(message)
            .setCustomTitle(titleLayout)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                ok?.let { activity.runOnUiThread(it) }
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                cancel?.let { activity.runOnUiThread(it) }
            }
            .show()
            .setCanceledOnTouchOutside(false)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(context: Context, message: Spanned, ok: Runnable?, cancel: Runnable? = null) {
        showConfirmation(context, MainApp.gs(R.string.confirmation), message, ok, cancel)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(context: Context, title: String, message: Spanned, ok: Runnable?, cancel: Runnable? = null) {
        val titleLayout = LayoutInflater.from(context).inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = title
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(context, R.style.AppTheme))
            .setMessage(message)
            .setCustomTitle(titleLayout)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                runOnUiThread(ok)
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                runOnUiThread(cancel)
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
            .setCanceledOnTouchOutside(false)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(context: Context, message: String, ok: Runnable?, cancel: Runnable? = null) {
        showConfirmation(context, MainApp.gs(R.string.confirmation), message, ok, cancel)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(context: Context, title: String, message: String, ok: Runnable?, cancel: Runnable? = null) {
        val titleLayout = LayoutInflater.from(context).inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = title
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(context, R.style.AppTheme))
            .setMessage(message)
            .setCustomTitle(titleLayout)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                runOnUiThread(ok)
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, _: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                runOnUiThread(cancel)
            }
            .show()
            .setCanceledOnTouchOutside(false)
    }

    @JvmStatic
    @JvmOverloads
    fun showConfirmation(context: Context, title: String, message: String, ok: DialogInterface.OnClickListener?, cancel: DialogInterface.OnClickListener? = null) {
        val titleLayout = LayoutInflater.from(context).inflate(R.layout.dialog_alert_custom, null)
        (titleLayout.findViewById<View>(R.id.alertdialog_title) as TextView).text = title
        (titleLayout.findViewById<View>(R.id.alertdialog_icon) as ImageView).setImageResource(R.drawable.ic_check_while_48dp)
        AlertDialog.Builder(ContextThemeWrapper(context, R.style.AppTheme))
            .setMessage(message)
            .setCustomTitle(titleLayout)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, which: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                ok?.onClick(dialog, which)
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, which: Int ->
                dialog.dismiss()
                SystemClock.sleep(100)
                cancel?.onClick(dialog, which)
            }
            .show()
            .setCanceledOnTouchOutside(false)
    }
}